apps/weathermap/weather_map.star

```diff
@@ -0,0 +1,359 @@
+    """Get the width a zoom level covers in kilometers and miles.
+
+    Args:
+        zoom_level: A zoom level.
+
+    Returns:
+        A tuple containing the values in kilometers and miles.
+    """
+    m_per_pixel = ZOOM_LEVEL_METERS_PER_PIXEL[zoom_level]
+    km = 256 * m_per_pixel // 1000
+    mi = km * 0.621371
+    return (km, mi)
+
+# A list of zoom levels to choose from
+#
+# This goes from "0" (earth) to "23" (2 cm per pixel)
+ZOOM_LEVELS = [
+    schema.Option(
+        display = "%d km / %d mi" % get_zoom_level_size(level),
+        value = "%d" % level,
+    )
+    for level in range(len(ZOOM_LEVEL_METERS_PER_PIXEL))
+]
+DEFAULT_ZOOM_LEVEL = ZOOM_LEVELS[7]
+
+DEFAULT_SNOW_OPTION = "1"
+DEFAULT_LOCATION = """
+{
+	"lat": "40.6781784",
+	"lng": "-73.9441579",
+	"description": "Brooklyn, NY, USA",
+	"locality": "Brooklyn",
+	"place_id": "ChIJCSF8lBZEwokRhngABHRcdoI",
+	"timezone": "America/New_York"
+}
+"""
+
+def render_frame(frame, image, opts):
+    """Render a frame of the weather layer animation.
+
+    Args:
+        frame: A frame object.
+        image: A binary image.
+        opts: A struct of rendering options.
+
+    Returns:
+        A definition of the frame to render.
+    """
+    time_format = getattr(opts, "time_format")
+    time_str = time.from_timestamp(int(frame["time"])).format(time_format)
+
+    zoom_level = int(getattr(opts, "zoom_level"))
+    (km, mi) = get_zoom_level_size(zoom_level)
+    unit_format = getattr(opts, "unit_format")
+    unit_str = "%d km" % km if unit_format == "km" else "%d mi" % mi
+
+    show_time = True if time_format != "off" else False
+    show_unit = True if unit_format != "off" else False
+
+    return render.Stack(children = [
+        render.Padding(
+            pad = (0, -16, 0, 0),
+            child = render.Image(
+                src = image,
+                width = 64,
+                height = 64,
+            ),
+        ),
+        render.Box(
+            height = 7,
+            color = "#0007",
+            child = render.Text(
+                content = time_str,
+                offset = -1,
+                color = "#fff",
+                font = "tom-thumb",
+            ),
+        ) if show_time else None,
+        render.Padding(
+            child = render.Box(
+                height = 7,
+                color = "#0007",
+                child = render.Text(
+                    content = unit_str,
+                    offset = -1,
+                    color = "#fff",
+                    font = "tom-thumb",
+                ),
+            ),
+            pad = (0, 32 - 7, 0, 0),
+        ) if show_unit else None,
+        render.Box(child = render.Circle(
+            diameter = 3,
+            color = "#f00a",
+        )),
+    ])
+
+def create_image_url(frame, opts):
+    return IMAGE_URL_LAYOUT.format(
+        host = getattr(opts, "host"),
+        path = frame["path"],
+        lat = getattr(opts, "location")["lat"],
```

Comment:
Would you mind truncating lat/lng? Mostly, we're concerned about leaking user data to third-party APIs. I think we could get away with truncating to two decimal places?? https://gis.stackexchange.com/questions/8650/measuring-accuracy-of-latitude-and-longitude